test/integration/daemonset/daemonset_test.go

```diff
@@ -393,3 +397,267 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
+	if err := wait.Poll(2*time.Second, 30*time.Second, func() (bool, error) {
+		pods := getPods(t, podClient, testLabels())
+		unavailablePods := 0
+		nodesToUpdatedPodCount := make(map[string]int)
+		for _, pod := range pods.Items {
+			if !metav1.IsControlledBy(&pod, ds) {
+				continue
+			}
+			podImage := pod.Spec.Containers[0].Image
+			if podImage == image {
+				nodesToUpdatedPodCount[pod.Spec.NodeName]++
+			}
+			if !podutil.IsPodAvailable(&pod, ds.Spec.MinReadySeconds, metav1.Now()) {
+				unavailablePods++
+			}
+		}
+		if unavailablePods > maxUnavailable {
+			return false, fmt.Errorf("number of unavailable pods %d is greater than maxUnavailable %d", unavailablePods, maxUnavailable)
+		}
+		// Make sure every pod on the node has been updated
+		nodeNames := schedulableNodes(c, ds, t)
+		for _, node := range nodeNames {
+			if nodesToUpdatedPodCount[node] == 0 {
+				return false, nil
+			}
+		}
+		return true, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// A DaemonSet should retry creating failed daemon pods
+func TestRetryCreatingFailedPod(t *testing.T) {
+	for _, strategy := range updateStrategies() {
+		server, closeFn, dc, informers, clientset := setup(t)
+		defer closeFn()
+		ns := framework.CreateTestingNamespace("retry-creating-failed-pod-test", server, t)
+		defer framework.DeleteTestingNamespace(ns, server, t)
+
+		dsClient := clientset.ExtensionsV1beta1().DaemonSets(ns.Name)
+		podClient := clientset.CoreV1().Pods(ns.Name)
+		nodeClient := clientset.CoreV1().Nodes()
+		podInformer := informers.Core().V1().Pods().Informer()
+		stopCh := make(chan struct{})
+		defer close(stopCh)
+		informers.Start(stopCh)
+		go dc.Run(5, stopCh)
+
+		dsName := "daemonset"
+		ds := newDaemonSet(dsName, ns.Name)
+		ds.Spec.UpdateStrategy = *strategy
+		ds, err := dsClient.Create(ds)
+		if err != nil {
+			t.Fatalf("failed to create daemonset %q: %v", dsName, err)
+		}
+
+		nodeName := "single-node"
+		node := newNode(nodeName, nil)
+		_, err = nodeClient.Create(node)
+		if err != nil {
+			t.Fatalf("failed to create node %q: %v", nodeName, err)
+		}
+
+		// Validate everything works correctly, include marking pods
+		// as ready and ensure they launch on every node
+		validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
+		validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 1, t)
+
+		// Set the pod's phase to "Failed"
+		pods := getPods(t, podClient, testLabels())
+		pod := updatePodStatus(t, podClient, pods.Items[0].Name, func(pod *v1.Pod) {
+			pod.Status.Phase = v1.PodFailed
+		})
+
+		// Validate everything works correctly again
+		validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
+		validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 1, t)
+
+		// Validate the failed pod is re-created
+		pods = getPods(t, podClient, testLabels())
+		pod = &pods.Items[0]
```

Comment:
I think we can make this check stronger, to ensure that the new replacement pod is on the same node.

Reply:
Good point! This test only creates one node, so i think the check whether the pod is on the same node is not necessary. However, for the second test which creates 2 nodes, it is necessary to check whether the pod is on the correct node.